velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp

```diff
@@ -96,17 +104,86 @@ std::unique_ptr<WriteFile> HdfsFileSystem::openFileForWrite(
   return std::make_unique<HdfsWriteFile>(impl_->hdfsClient(), path);
 }
 
-bool HdfsFileSystem::isHdfsFile(const std::string_view filename) {
-  return filename.find(kScheme) == 0;
+bool HdfsFileSystem::isHdfsFile(const std::string_view filePath) {
+  return filePath.find(kScheme) == 0;
+}
+
+/**
+ * Get hdfs endpoint from config. This is applicable to the case that only one
+ * hdfs endpoint will be used.
+ */
+HdfsServiceEndpoint HdfsFileSystem::getServiceEndpoint(const Config* config) {
+  auto hdfsHost = config->get("hive.hdfs.host");
+  VELOX_CHECK(
+      hdfsHost.hasValue(),
+      "hdfsHost is empty, configuration missing for hdfs host");
+  auto hdfsPort = config->get("hive.hdfs.port");
+  VELOX_CHECK(
+      hdfsPort.hasValue(),
+      "hdfsPort is empty, configuration missing for hdfs port");
+  HdfsServiceEndpoint endpoint{*hdfsHost, *hdfsPort};
+  return endpoint;
+}
+
+/**
+ * Get hdfs endpoint from a given file path, instead of getting a fixed one from
+ * configuration.
+ */
+HdfsServiceEndpoint HdfsFileSystem::getServiceEndpoint(
+    const std::string_view filePath) {
+  auto index1 = filePath.find('/', kScheme.size() + 1);
+  std::string hdfsIdentity{
+      filePath.data(), kScheme.size(), index1 - kScheme.size()};
+  VELOX_CHECK(
+      !hdfsIdentity.empty(),
+      "hdfsIdentity is empty, expect hdfs endpoint host[:port] is contained in file path");
+  auto index2 = hdfsIdentity.find(':', 0);
+  // In HDFS HA mode, the hdfsIdentity is a nameservice ID with no port.
+  if (index2 == std::string::npos) {
+    HdfsServiceEndpoint endpoint{hdfsIdentity, ""};
+    return endpoint;
+  }
+  std::string host{hdfsIdentity.data(), 0, index2};
+  std::string port{
+      hdfsIdentity.data(), index2 + 1, hdfsIdentity.size() - index2 - 1};
+  HdfsServiceEndpoint endpoint{host, port};
+  return endpoint;
 }
 
-static std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
-    filesystemGenerator = [](std::shared_ptr<const Config> properties) {
-      static std::shared_ptr<FileSystem> filesystem;
-      folly::call_once(hdfsInitiationFlag, [&properties]() {
-        filesystem = std::make_shared<HdfsFileSystem>(properties);
-      });
-      return filesystem;
+static std::function<std::shared_ptr<FileSystem>(
+    std::shared_ptr<const Config>,
+    std::string_view)>
+    filesystemGenerator = [](std::shared_ptr<const Config> properties,
+                             std::string_view filePath) {
+      static folly::ConcurrentHashMap<std::string, std::shared_ptr<FileSystem>>
+          filesystems;
+      static folly::
+          ConcurrentHashMap<std::string, std::shared_ptr<folly::once_flag>>
+              hdfsInitiationFlags;
+      auto endpoint = HdfsFileSystem::getServiceEndpoint(filePath);
```

Comment:
I have seen the case where the filePath is specified without hdfs://serverAddress:port. In this case, should we fallback to the `hive.hdfs.host` & `hive.hdfs.port` from the config file?

Reply:
Thanks for your comment!
Yes, it's a good idea. I just added the fallback support in the new commit. For file path given like `hdfs:///your_hdfs_file` with no host/port contained, it will fall back to find endpoint info from config.